### ISSUE TITLE
Analyze function pointer transmutes

### ISSUE THREAD

**Issue by jswrenn** [rust-lang/project-safe-transmute#21]
This was heavily-requested at RustConf. This analysis probably consists of a ABI equality check, and a pairwise, conventional transmutability analysis of the arguments and return types. 

For example, this should be rejected:
```rust
extern "C" fn c() {
    println!("c");
}

fn main() {
    let src = c as extern "C" fn();
    let dst: fn() -> () = unsafe { core::mem::transmute(src)};
    dst()
}
```

**Comment by Lokathor** [comment]
Don't different types pass/return in different registers sometimes? Eg, f32 and u32 transmute to each other, but don't pass in the same register on all architectures.

I think fn pointer transmute is far more restricted than normal transmute.

**Comment by jswrenn** [comment]
Yeah, that's why we also would have to check the ABI/calling convention. I think we can use [`test_abi_eq`](https://doc.rust-lang.org/beta/nightly-rustc/rustc_passes/abi_test/fn.test_abi_eq.html) for that.
